### PR TITLE
#1610 tags are lost upon updating discussion

### LINKED ISSUE #1610: All tags are lost upon updating a discussion
**To Reproduce**
1. Create a discussion with a few tags
2. Edit the discussion
3. Added tags are no longer there

**Expected behavior**
Tags need to be listed in the tags field during edits

### PR DESCRIPTION
**Pull request by sadrul**
### All Submissions:

* [x] Have you followed the [Pull Requests Guidelines](https://github.com/buddyboss/buddyboss-platform/wiki/Submitting-Pull-Requests#pull-request-guidelines)?
* [x] Does your code follow the [WordPress' Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../pulls) for the same update/change?

### Changes proposed in this Pull Request:

Fixes #1610

### How to test the changes in this Pull Request:

1. Create a discussion with a few tags
2. Edit the discussion
3. See the tags in tags section

### Proof Screenshots or Video

https://www.loom.com/share/0e45a53eff144c4eaaecfcb24ba920e8

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [x] Have you successfully run tests with your changes locally?

### Related Theme PR:
buddyboss/buddyboss-theme#1136

### CODE REVIEW

File: src/bp-forums/templates/default/bbpress/form-topic.php [hunk 1]
```diff
@@ -90,9 +90,26 @@
...
 
+						<?php
+						$get_topic_id = bbp_get_topic_id();
+						$get_the_tags = isset($get_topic_id) && !empty($get_topic_id) ? bbp_get_topic_tag_names($get_topic_id) : '';
+						?>
+
 						<p>
-							<input type="hidden" value="" name="bbp_topic_tags" class="bbp_topic_tags" id="bbp_topic_tags" >
-							<select name="bbp_topic_tags_dropdown[]" class="bbp_topic_tags_dropdown" id="bbp_topic_tags_dropdown" placeholder="<?php esc_html_e( 'Type one or more tag, comma separated', 'buddyboss' ); ?>" autocomplete="off" multiple="multiple" style="width: 100%" tabindex="<?php bbp_tab_index(); ?>"></select>
+							<input type="hidden" value="<?php echo $get_the_tags; ?>" name="bbp_topic_tags" class="bbp_topic_tags" id="bbp_topic_tags" >
```
Comment (dipskakadiya) [inline src/bp-forums/templates/default/bbpress/form-topic.php:99]: @sadrul Can we use escaping function for `get_the_tags`
```suggestion
							<input type="hidden" value="<?php echo esc_attr( $get_the_tags ); ?>" name="bbp_topic_tags" class="bbp_topic_tags" id="bbp_topic_tags" >
```

File: src/bp-forums/templates/default/bbpress/form-topic.php [hunk 2]
```diff
@@ -90,9 +90,26 @@
...
+							<select name="bbp_topic_tags_dropdown[]" class="bbp_topic_tags_dropdown" id="bbp_topic_tags_dropdown" placeholder="<?php esc_html_e( 'Type one or more tag, comma separated', 'buddyboss' ); ?>" autocomplete="off" multiple="multiple" style="width: 100%" tabindex="<?php bbp_tab_index(); ?>">
+								<?php
+								if( isset($get_the_tags) && !empty($get_the_tags) ){
+									$get_the_tags = explode(',', $get_the_tags);
+									foreach($get_the_tags as $single_tag){
+										$single_tag = trim($single_tag);
+										?>
+										<option selected="selected" value="<?php echo $single_tag; ?>"><?php echo $single_tag; ?></option>
```
Comment (dipskakadiya) [inline src/bp-forums/templates/default/bbpress/form-topic.php:107]: @sadrul Missing escaping function here too.

**Review by dipskakadiya — CHANGES_REQUESTED** [review]
@sadrul Can we fix minor escaping missing and also resolve conflict.

**Comment by sadrul** [comment]
@dipskakadiya I updated accordingly. Linked buddyboss-theme PR
https://github.com/buddyboss/buddyboss-theme/pull/1136

File: src/bp-forums/templates/default/bbpress/form-reply.php [hunk 3]
```diff
@@ -65,9 +65,26 @@
...
 
+						<?php
+						$get_topic_id = bbp_get_topic_id();
+						$get_the_tags = isset( $get_topic_id ) && !empty( $get_topic_id ) ? bbp_get_topic_tag_names( $get_topic_id ) : '';
+						?>
+
 						<p>
 							<input type="hidden" value="" name="bbp_topic_tags" class="bbp_topic_tags" id="bbp_topic_tags" >
-							<select name="bbp_topic_tags_dropdown[]" class="bbp_topic_tags_dropdown" id="bbp_topic_tags_dropdown" placeholder="<?php esc_html_e( 'Type one or more tag, comma separated', 'buddyboss' ); ?>" autocomplete="off" multiple="multiple" style="width: 100%" tabindex="<?php bbp_tab_index(); ?>"></select>
+							<select name="bbp_topic_tags_dropdown[]" class="bbp_topic_tags_dropdown" id="bbp_topic_tags_dropdown" placeholder="<?php esc_html_e( 'Type one or more tag, comma separated', 'buddyboss-theme' ); ?>" autocomplete="off" multiple="multiple" style="width: 100%" tabindex="<?php bbp_tab_index(); ?>">
```
Comment (chetansatasiya) [inline src/bp-forums/templates/default/bbpress/form-reply.php:75]:
```suggestion
							<select name="bbp_topic_tags_dropdown[]" class="bbp_topic_tags_dropdown" id="bbp_topic_tags_dropdown" placeholder="<?php esc_html_e( 'Type one or more tag, comma separated', 'buddyboss' ); ?>" autocomplete="off" multiple="multiple" style="width: 100%" tabindex="<?php bbp_tab_index(); ?>">
```

**Review by chetansatasiya — CHANGES_REQUESTED** [review]
Need requested changes

File: src/bp-forums/replies/functions.php [hunk 4]
```diff
@@ -738,6 +738,25 @@ function bbp_edit_reply_handler( $action = '' ) {
 	// Just in time manipulation of reply terms before being edited
 	$terms = apply_filters( 'bbp_edit_reply_pre_set_terms', $terms, $topic_id, $reply_id );
 
+	// remove deleted terms
+	$existing_terms = bbp_get_topic_tag_names( $topic_id );
+
+	if ( !empty( $existing_terms ) ) {
```
Comment (KartikSuthar) [inline src/bp-forums/replies/functions.php:744]:
```suggestion
	if ( ! empty( $existing_terms ) ) {
```

File: src/bp-forums/replies/functions.php [hunk 5]
```diff
@@ -738,6 +738,25 @@ function bbp_edit_reply_handler( $action = '' ) {
...
+		$deleted_terms = [];
+		$existing_terms = explode( ',', $existing_terms );
+		$existing_terms = array_map( function ( $single ) {
+			return trim( $single );
+		}, $existing_terms );
+
+		if ( !empty( $terms ) ) {
```
Comment (KartikSuthar) [inline src/bp-forums/replies/functions.php:751]:
```suggestion
		if ( ! empty( $terms ) ) {
```

File: src/bp-forums/replies/functions.php [hunk 6]
```diff
@@ -738,6 +738,25 @@ function bbp_edit_reply_handler( $action = '' ) {
...
+			$terms_array = explode( ',', $terms );
+			$deleted_terms = array_diff( $existing_terms, $terms_array );
+		}
+		if ( !empty( $deleted_terms ) ) {
```
Comment (KartikSuthar) [inline src/bp-forums/replies/functions.php:755]:
```suggestion
		if ( ! empty( $deleted_terms ) ) {
```

File: src/bp-forums/templates/default/bbpress/form-reply.php [hunk 7]
```diff
@@ -65,9 +65,26 @@
 
 						<?php do_action( 'bbp_theme_before_reply_form_tags' ); ?>
 
+						<?php
+						$get_topic_id = bbp_get_topic_id();
+						$get_the_tags = isset( $get_topic_id ) && !empty( $get_topic_id ) ? bbp_get_topic_tag_names( $get_topic_id ) : '';
```
Comment (KartikSuthar) [inline src/bp-forums/templates/default/bbpress/form-reply.php:70]:
```suggestion
						$get_the_tags = isset( $get_topic_id ) && ! empty( $get_topic_id ) ? bbp_get_topic_tag_names( $get_topic_id ) : '';
```

File: src/bp-forums/templates/default/bbpress/form-reply.php [hunk 8]
```diff
@@ -65,9 +65,26 @@
...
+						?>
+
 						<p>
 							<input type="hidden" value="" name="bbp_topic_tags" class="bbp_topic_tags" id="bbp_topic_tags" >
-							<select name="bbp_topic_tags_dropdown[]" class="bbp_topic_tags_dropdown" id="bbp_topic_tags_dropdown" placeholder="<?php esc_html_e( 'Type one or more tag, comma separated', 'buddyboss' ); ?>" autocomplete="off" multiple="multiple" style="width: 100%" tabindex="<?php bbp_tab_index(); ?>"></select>
+							<select name="bbp_topic_tags_dropdown[]" class="bbp_topic_tags_dropdown" id="bbp_topic_tags_dropdown" placeholder="<?php esc_html_e( 'Type one or more tag, comma separated', 'buddyboss-theme' ); ?>" autocomplete="off" multiple="multiple" style="width: 100%" tabindex="<?php bbp_tab_index(); ?>">
+								<?php
+								if( !empty( $get_the_tags ) ) {
```
Comment (KartikSuthar) [inline src/bp-forums/templates/default/bbpress/form-reply.php:77]:
```suggestion
								if ( ! empty( $get_the_tags ) ) {
```

File: src/bp-forums/templates/default/bbpress/form-reply.php [hunk 9]
```diff
@@ -65,9 +65,26 @@
...
+									$get_the_tags = explode( ',', $get_the_tags );
+									foreach( $get_the_tags as $single_tag ) {
```
Comment (KartikSuthar) [inline src/bp-forums/templates/default/bbpress/form-reply.php:79]:
```suggestion
									foreach ( $get_the_tags as $single_tag ) {
```

File: src/bp-forums/templates/default/bbpress/form-topic.php [hunk 10]
```diff
@@ -90,9 +90,26 @@
 
 						<?php do_action( 'bbp_theme_before_topic_form_tags' ); ?>
 
+						<?php
+						$get_topic_id = bbp_get_topic_id();
+						$get_the_tags = isset( $get_topic_id ) && !empty( $get_topic_id ) ? bbp_get_topic_tag_names( $get_topic_id ) : '';
```
Comment (KartikSuthar) [inline src/bp-forums/templates/default/bbpress/form-topic.php:95]:
```suggestion
						$get_the_tags = isset( $get_topic_id ) && ! empty( $get_topic_id ) ? bbp_get_topic_tag_names( $get_topic_id ) : '';
```

File: src/bp-forums/templates/default/bbpress/form-topic.php [hunk 11]
```diff
@@ -90,9 +90,26 @@
...
+						?>
+
 						<p>
-							<input type="hidden" value="" name="bbp_topic_tags" class="bbp_topic_tags" id="bbp_topic_tags" >
-							<select name="bbp_topic_tags_dropdown[]" class="bbp_topic_tags_dropdown" id="bbp_topic_tags_dropdown" placeholder="<?php esc_html_e( 'Type one or more tag, comma separated', 'buddyboss' ); ?>" autocomplete="off" multiple="multiple" style="width: 100%" tabindex="<?php bbp_tab_index(); ?>"></select>
+							<input type="hidden" value="<?php echo esc_attr( $get_the_tags ); ?>" name="bbp_topic_tags" class="bbp_topic_tags" id="bbp_topic_tags" >
+							<select name="bbp_topic_tags_dropdown[]" class="bbp_topic_tags_dropdown" id="bbp_topic_tags_dropdown" placeholder="<?php esc_html_e( 'Type one or more tag, comma separated', 'buddyboss' ); ?>" autocomplete="off" multiple="multiple" style="width: 100%" tabindex="<?php bbp_tab_index(); ?>">
+								<?php
+								if( !empty( $get_the_tags ) ) {
```
Comment (KartikSuthar) [inline src/bp-forums/templates/default/bbpress/form-topic.php:102]:
```suggestion
								if ( ! empty( $get_the_tags ) ) {
```

File: src/bp-forums/templates/default/bbpress/form-topic.php [hunk 12]
```diff
@@ -90,9 +90,26 @@
...
+									$get_the_tags = explode( ',', $get_the_tags );
+									foreach( $get_the_tags as $single_tag ) {
```
Comment (KartikSuthar) [inline src/bp-forums/templates/default/bbpress/form-topic.php:104]:
```suggestion
									foreach ( $get_the_tags as $single_tag ) {
```

File: src/bp-forums/replies/functions.php [hunk 13]
```diff
@@ -738,6 +738,25 @@ function bbp_edit_reply_handler( $action = '' ) {
 	// Just in time manipulation of reply terms before being edited
 	$terms = apply_filters( 'bbp_edit_reply_pre_set_terms', $terms, $topic_id, $reply_id );
 
+	// remove deleted terms
+	$existing_terms = bbp_get_topic_tag_names( $topic_id );
+
+	if ( !empty( $existing_terms ) ) {
+		$deleted_terms = [];
```
Comment (KartikSuthar) [inline src/bp-forums/replies/functions.php:745]:
```suggestion
		$deleted_terms = array();
```

**Review by KartikSuthar — CHANGES_REQUESTED** [review]

**Comment by sadrul** [comment]
@KartikSuthar @chetansatasiya done with change requests

**Review by KartikSuthar — APPROVED** [review]
@sadrul Merged to dev and pass to QA.

**Comment by sadrul** [comment]
@mohit-buddyboss @sumitbuddyboss Merged into dev and ready for testing

**Comment by mohit-buddyboss** [comment]
@sadrul 

Tested using dev branch of the platform. Tested this issue for other themes like 2020,2021 for older discussion as well as by creating new discussions. 

Video: https://loom.com/share/f4e1eb049e7142688e7517ac10c6cbfe

Related issue is resolve now and PR is ready for release.

cc @sourabhmatolia @KartikSuthar 

Related Theme PR:
buddyboss/buddyboss-theme#1136

**Comment by JetalDobariya** [comment]
Hello

I have tested this PR in QA Approved branch , Tags are appear in the Discussion while we Creating/Editing.
But I facing some issues like Tags are not getting deleted while we edit second time.

Here, I drop videos link for your reference, please check and let me know if you have any Query!
https://www.loom.com/share/2d87a0b69f3543ed995e0cb340b1d544

Thanks!

**Comment by sadrul** [comment]
@JetalDobariya I fixed the issue you reported in above video (regarding remove tag). 
But as we discussed another issue that new tag does not add when tag exists with same slug https://i.imgur.com/sU8RO4b.png
I created a new issue for this bug https://github.com/buddyboss/buddyboss-platform/issues/2858

**Comment by JetalDobariya** [comment]
Hello,

I did verify this PR's issue has been done in QA Approved branch, now tags are getting deleted.
See video : https://www.loom.com/share/0b8083d162e14cc989fcd3848f2c0a9f

Thanks!
Jetal!
CC : @sadrul

**Pull request merged**